alns/accept/GreatDeluge.py

```diff
@@ -0,0 +1,141 @@
+import math
+from typing import Optional
+
+from alns.accept.AcceptanceCriterion import AcceptanceCriterion
+
+
+class GreatDeluge(AcceptanceCriterion):
+    """
+    The Great Deluge (GD) criterion accepts solutions if the candidate solution
+    has value lower than a threshold (originally called the water level). The
+    initial threshold is computed as
+
+    ``threshold = alpha * initial.objective()``
+
+    where `initial` is the initial solution passed-in to ALNS, inferred
+    from the best solution at the first iteration.
+
+    There are two variants of the GD criterion: 1) linear and 2) non-linear,
+    proposed in [2] and [3], respectively. The main difference between these
+    variants is the threshold updating scheme, see the `_update` method for
+    more details. Moreover, in the non-linear variant, candidate solutions that
+    improve the current solution are always accepted.
+
+    The implementation is based on the description in [1].
+
+    Parameters
+    ----------
+    alpha
+        The factor to compute the initial threshold
+    beta
+        A factor used for updating the threshold
+    gamma
+        A factor used for updating the threshold (only in non-linear GD)
+    delta
+        A factor used for updating the threshold (only in non-linear GD)
+    method
+        The updating method, one of {'linear', 'non-linear'}. Default 'linear'.
+
+    References
+    ----------
+    [1]: Santini, A., Ropke, S. & Hvattum, L.M. A comparison of acceptance
+         criteria for the adaptive large neighbourhood search metaheuristic.
+         *Journal of Heuristics* (2018) 24 (5): 783–815.
+    [2]: Dueck, G. New optimization heuristics: The great deluge algorithm and
+         the record-to-record travel. *Journal of Computational Physics* (1993)
+         104 (1): 86-92.
+    [3]: Landa-Silva, D., & Obit, J. H. Great deluge with non-linear decay rate
+         for solving course timetabling problems. *4th international IEEE
+         conference intelligent systems* (2008) Vol. 1: 8-11.
+    """
+
+    def __init__(
+        self,
+        alpha: float,
+        beta: float,
+        gamma: Optional[float] = None,
+        delta: Optional[float] = None,
+        method: str = "linear",
+    ):
+        if alpha <= 1:
+            raise ValueError("Alpha must be larger than 1.")
+
+        if beta <= 0 or beta >= 1:
+            raise ValueError("Beta must be in (0, 1).")
+
+        if (
+            gamma is None or delta is None or gamma <= 0 or delta <= 0
+        ) and method == "non-linear":
+            raise ValueError("Gamma and delta must be non-negative.")
```

Comment:
Question: these parameters are optional with the linear method, right?

Refactor: I think we should split this class up. Every member method branches on whether it is the linear/non-linear method. There's very benefit we're getting from having one GreatDeluge class.

I propose we do a `LinearGreatDeluge` and `NonLinearGreatDeluge` class. Those should then be rather straightforward.

Reply:
Answer: that's correct.

Refactoring is a good idea. I prefer to name it `GreatDeluge`, which is how the criterion's name as known in the literature.

Reply:
Agreed! I was only thinking from the code, but if the literature calls it GreatDeluge we should probably do so too.

What's the non-linear variant called? Does that have a name other than non-linear great deluge?

Reply:
The [original paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.723.2759&rep=rep1&type=pdf) calls it "Non-Linear Great Deluge". 

I haven't looked for enough great deluge papers to know if there are other names, but Santini et al. (2018) call it NLGD as well and searching for non-linear great deluge on [Google scholar ](
https://scholar.google.com/scholar?hl=nl&as_sdt=0%2C5&q=non+linear+great+deluge&btnG=)seems to give enough matches.